src/github.com/mozilla-services/pushgo/simplepush/dynamo_store.go

```diff
@@ -0,0 +1,507 @@
+	if len(tableSet) == 0 {
+		return false
+	}
+	for _, t := range tableSet {
+		if t == table {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *DynamoDBStore) getNow(t int64) string {
+	if t == 0 {
+		t = time.Now().UTC().Unix()
+	}
+	return strconv.FormatInt(t, 10)
+}
+
+// Init initializes the memcached adapter with the given configuration.
+// Implements HasConfigStruct.Init().
+func (s *DynamoDBStore) Init(app *Application, config interface{}) (err error) {
+	var ok bool
+	conf := config.(*DynamoDBConf)
+	s.logger = app.Logger()
+
+	// create the server
+	s.region, ok = aws.Regions[conf.Region]
+	if !ok {
+		s.logger.Panic("dynamodb", "Invalid region specified", nil)
+		return ErrInvalidRegion
+	}
+	auth, err := aws.GetAuth(conf.Access, conf.Secret,
+		"", time.Now().Add(s.statusTimeout))
+	if err != nil {
+		s.logger.Panic("dynamodb", "Could not log into dynamodb",
+			LogFields{"error": err.Error()})
+		return
+	}
+
+	if s.server == nil {
+		s.server = &dynamodb.Server{auth, s.region}
+	}
+	tables, err := s.server.ListTables()
+	if err != nil {
+		s.logger.Panic("dynamodb", "Could not query dynamodb",
+			LogFields{"error": err.Error()})
+		return
+	}
+	s.readProv = conf.ReadProv
+	s.writeProv = conf.WriteProv
+	s.maxChannels = conf.MaxChannels
+	s.tablename = conf.TableName
+
+	// check if the table exists
+	if !s.containsTable(tables, conf.TableName) {
+		if s.logger.ShouldLog(INFO) {
+			s.logger.Info("dynamodb", "creating dynamodb table...", nil)
+		}
+		if err = s.createTable(); err != nil {
+			s.logger.Panic("dynamodb", "Could not create table",
+				LogFields{"error": err.Error()})
+			return
+		}
+	}
+	s.pk = dynamodb.PrimaryKey{dynamodb.NewStringAttribute(UAID_LABEL, ""),
+		dynamodb.NewStringAttribute(CHID_LABEL, "")}
+	s.table = s.server.NewTable(conf.TableName, s.pk)
+	return nil
+}
+
+// CanStore indicates whether the specified number of channel registrations
+// are allowed per client. Implements Store.CanStore().
+func (s *DynamoDBStore) CanStore(channels int) bool {
+	return channels <= s.maxChannels
+}
+
+// Close closes the connection pool and unblocks all pending operations with
+// errors. Safe to call multiple times. Implements Store.Close().
+func (s *DynamoDBStore) Close() (err error) {
+	return
+}
+
+// KeyToIDs extracts the hex-encoded device and channel IDs from a user-
+// readable primary key. Implements Store.KeyToIDs().
+func (*DynamoDBStore) KeyToIDs(key string) (uaid, chid string, err error) {
+	items := strings.SplitN(key, ".", 2)
+	if len(items) < 2 {
+		return "", "", nil
+	}
+	return items[0], items[1], nil
+}
+
+// IDsToKey generates a user-readable primary key from a (device ID, channel
+// ID) tuple. The primary key is encoded in the push endpoint URI. Implements
+// Store.IDsToKey().
+func (*DynamoDBStore) IDsToKey(uaid, chid string) (string, error) {
+	if len(uaid) == 0 || len(chid) == 0 {
+		return "", nil
+	}
+	return fmt.Sprintf("%s.%s", uaid, chid), nil
```

Comment:
Sprintf is the slowest way to join strings. join, concat or basic adding them together is much faster.

https://groups.google.com/forum/#!msg/golang-nuts/uIhmp50Agtw/A47Kpgrbb2kJ